File: adventure-engine/backend/app/api/routes/discovery.py
from fastapi import APIRouter, HTTPException

from app.schemas.discovery import DiscoveryRequestSchema, DiscoveryResponse
from app.services.discovery.engine import discover
from app.services.discovery.serpapi_client import SerpApiNotConfiguredError
from app.services.discovery.types import DiscoveryRequest

router = APIRouter(prefix="/discover", tags=["discovery"])


@router.post("/", response_model=DiscoveryResponse)
def discover_activities(body: DiscoveryRequestSchema) -> DiscoveryResponse:
    """Runs the Activity Discovery Engine (SerpAPI-based, see
    documentation/activity_discovery_engine.md) and returns the 7 named
    recommendation buckets plus a chained route through "Best Overall".
    """
    request = DiscoveryRequest(
        latitude=body.latitude,
        longitude=body.longitude,
        location_label=body.location_label,
        interests=body.interests or [],
        free_text=body.free_text,
        max_budget=body.max_budget,
    )
    try:
        result = discover(request)
    except SerpApiNotConfiguredError:
        raise HTTPException(status_code=503, detail="SerpAPI is not configured (SERPAPI_KEY missing).")

    return DiscoveryResponse.from_result(result)

File: adventure-engine/frontend/src/services/discovery.ts
import { apiPost } from './api';
import type { DiscoveryResponse } from './types';

export interface DiscoveryRequestParams {
  latitude: number;
  longitude: number;
  locationLabel: string;
  interests?: string[];
  freeText?: string;
  maxBudget?: number;
}

export function discoverActivities(params: DiscoveryRequestParams): Promise<DiscoveryResponse> {
  return apiPost<DiscoveryResponse>('/discover/', {
    latitude: params.latitude,
    longitude: params.longitude,
    location_label: params.locationLabel,
    interests: params.interests,
    free_text: params.freeText,
    max_budget: params.maxBudget,
  });
}
